File: android/app/src/main/java/com/homeboy/app/ai/TagSuggestionService.kt
package com.homeboy.app.ai

import android.content.Context
import com.homeboy.app.api.HBTag
import java.io.File

/**
 * Turns an item's name + description into tag suggestions using the on-device LLM. Output is
 * split into tags that already exist (so we can select them directly) and novel suggestions the
 * user can create with one tap. Returns null whenever generation is unavailable, so callers
 * simply show nothing.
 */
object TagSuggestionService {

    data class Suggestions(
        val existing: List<HBTag>,
        val novel: List<String>,
        val modelLabel: String,
        val backend: AiBackend
    ) {
        val isEmpty: Boolean get() = existing.isEmpty() && novel.isEmpty()
    }

    suspend fun suggest(
        context: Context,
        modelId: String,
        modelName: String,
        modelFile: File,
        name: String,
        description: String,
        existingTags: List<HBTag>,
        unloadMinutes: Int,
        preferred: AiBackend? = null
    ): Suggestions? {
        if (name.isBlank()) return null
        if (!LlmEngineManager.ensureLoaded(context, modelId, modelFile, preferred)) return null

        val (system, user) = buildMessages(name, description, existingTags)
        val raw = LlmEngineManager.generateChat(system, user) ?: run {
            LlmEngineManager.scheduleUnload(unloadMinutes)
            return null
        }
        LlmEngineManager.scheduleUnload(unloadMinutes)

        val names = parse(raw)
        // Reconcile the model's suggestions against the library HERE, not in the prompt: a
        // suggestion that maps to an existing tag is shown as already-existing (no "+"), the rest
        // are offered as new tags to create. Matching is plural/case-insensitive so a generated
        // "Spices" still takes precedence over an existing "Spice".
        val existingBySingular = existingTags.associateBy { singular(it.name) }
        val matched = LinkedHashSet<HBTag>()
        val novel = LinkedHashSet<String>()
        for (n in names) {
            val tag = existingBySingular[singular(n)]
            // Existing tags keep their stored casing; new tags get title-cased for display.
            if (tag != null) matched.add(tag) else novel.add(titleCase(n))
        }
        val backend = (LlmEngineManager.state.value as? LlmEngineManager.State.Ready)?.backend
            ?: AiBackend.CPU
        return Suggestions(matched.toList(), novel.toList(), modelName, backend)
    }

    /**
     * Returns (systemPrompt, userMessage) for a chat-template-aware model call.
     *
     * Design note: the model's ONE job is to generate the best tags for the item itself. We
     * deliberately do NOT ask it to pick from the existing-tag list — small instruct models
     * (Qwen3 1.7B) read "reuse an existing tag whenever one fits" as "classify into this list" and
     * will force-fit unrelated tags (tagging "black pepper" as "Pesticides") instead of inventing
     * the obvious "Spices"/"Cooking". Existing-vs-novel is reconciled in code afterwards (see
     * [suggest]); the library is passed here only as a soft spelling hint so a generated tag aligns
     * to an existing one's wording. A worked example anchors the output format, and `/no_think`
     * suppresses Qwen3 chain-of-thought.
     */
    private fun buildMessages(
        name: String,
        description: String,
        existingTags: List<HBTag>
    ): Pair<String, String> {
        val tagList = existingTags.take(40).joinToString(", ") { it.name }
        val system = buildString {
            append("/no_think\n")
            append("You suggest organizing tags for an item in a home inventory. ")
            append("Reply with ONLY a comma-separated list of 4 to 6 short tags (1-2 words each) and nothing else. ")
            append("Pick tags that genuinely describe the item — its kind, category, use, or where it is kept. ")
            append("Always invent tags that fit the item; never force an unrelated tag just to reuse one. ")
            append("Example — \"cordless drill\": Tools, Power Tools, Garage, Hardware, DIY.")
            if (tagList.isNotBlank()) {
                append("\nIf a tag you choose means the same as one already in the library, copy its exact spelling: ")
                append(tagList).append('.')
            }
        }
        val user = buildString {
            append("Item: ").append(name)
            if (description.isNotBlank()) append("\nDescription: ").append(description)
            append("\nTags:")
        }
        return system to user
    }

    /** Pull a clean tag list out of the model's free-form reply. */
    private fun parse(raw: String): List<String> {
        // Strip <think>...</think> blocks produced by reasoning models that ignore /no_think.
        val stripped = stripThinking(raw)
        val body = if (stripped.contains("Tags:")) stripped.substringAfterLast("Tags:") else stripped
        val firstLine = body.lineSequence().map { it.trim() }.firstOrNull { it.isNotBlank() }
            ?: return emptyList()
        return firstLine.split(',', ';')
            .map { it.trim().trim('-', '*', '.', '"', '\'', '#', '`', ' ') }
            .filter { it.length in 2..30 && it.split(' ').size <= 3 }
            .distinctBy { it.lowercase() }
            .take(6)
    }

    private fun stripThinking(raw: String): String {
        var s = raw
        while (true) {
            val start = s.indexOf("<think>")
            if (start < 0) break
            val end = s.indexOf("</think>", start)
            s = if (end < 0) s.substring(0, start)
                else s.substring(0, start) + s.substring(end + 8)
        }
        return s.trim()
    }

    /**
     * Capitalise a model-proposed tag for display, e.g. "home improvement" -> "Home Improvement".
     * Only all-lowercase words are touched, so deliberate casing like "USB" or "iPhone" survives.
     */
    private fun titleCase(s: String): String =
        s.split(' ').joinToString(" ") { word ->
            if (word.isNotEmpty() && word.none { it.isUpperCase() })
                word.replaceFirstChar { it.uppercase() }
            else word
        }

    /**
     * Fold a tag name to a match key: lowercase, trimmed, collapsed spaces, with a simple plural
     * suffix removed so "Spices" matches an existing "Spice" and "Batteries" matches "Battery".
     * Sibilant "-es" plurals (boxes, glasses) aren't folded — rare among inventory tags, and the
     * worst case is just offering a near-duplicate as a new tag rather than a wrong match.
     */
    private fun singular(s: String): String {
        val n = s.lowercase().trim().replace(Regex("\\s+"), " ")
        return when {
            n.endsWith("ies") && n.length > 4 -> n.dropLast(3) + "y"
            n.endsWith("ss") -> n
            n.endsWith("s") && n.length > 3 -> n.dropLast(1)
            else -> n
        }
    }
}

File: TagSuggestionService.swift
import Foundation
import FoundationModels

/// Result of an AI tag suggestion: existing tags to offer for quick-select, plus
/// novel tag names (title-cased) that aren't in the library yet.
struct TagSuggestions: Sendable {
    var matchedIds: [String]
    var novel: [String]

    var isEmpty: Bool { matchedIds.isEmpty && novel.isEmpty }
}

/// Provider-aware tag suggestion.
/// - `apple` — Apple Foundation Models (`SystemLanguageModel`, Neural Engine). Default.
/// - `gguf` — llama.cpp GGUF generation models. Wired in Phase 2.
///
/// Suggestion post-processing (reconciliation, plural folding, title-casing, thinking
/// stripping) is ported from the Android `TagSuggestionService` so both platforms
/// behave the same. See the `llm-integration-patterns` memory.
@MainActor
final class TagSuggestionService {

    /// Returns suggestions, or `nil` if the chosen provider is unavailable.
    func suggest(name: String,
                 description: String,
                 existing: [HBTag],
                 provider: LLMProvider) async -> TagSuggestions? {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 3 else { return nil }
        switch provider {
        case .apple:
            return await suggestApple(name: trimmed, description: description, existing: existing)
        case .gguf:
            return await suggestGGUF(name: trimmed, description: description, existing: existing)
        }
    }

    // MARK: - Apple Foundation Models

    private func suggestApple(name: String, description: String, existing: [HBTag]) async -> TagSuggestions? {
        guard SystemLanguageModel.default.isAvailable else { return nil }
        do {
            let session = LanguageModelSession()
            let response = try await session.respond(
                to: tagInstructions(noThink: false) + "\n\n" + tagUser(name: name, description: description))
            return reconcile(raw: "\(response.content)", existing: existing)
        } catch {
            return nil
        }
    }

    // MARK: - GGUF (llama.cpp) via the shared GenerationEngine

    private func suggestGGUF(name: String, description: String, existing: [HBTag]) async -> TagSuggestions? {
        // /no_think suppresses Qwen3 reasoning; reconcile() also strips any <think> that leaks.
        guard let raw = await GenerationEngine.shared.generate(
            system: tagInstructions(noThink: true),
            user: tagUser(name: name, description: description)) else { return nil }
        return reconcile(raw: raw, existing: existing)
    }

    // We deliberately do NOT feed the existing tag list as the choice set — the model
    // would force-fit unrelated tags. It suggests freely; we reconcile in code.
    private func tagInstructions(noThink: Bool) -> String {
        let base = """
        You label home-inventory items with short tags. Reply with ONLY a comma-separated \
        list of 3 to 6 tags, each 1 or 2 words, and nothing else. Pick tags that genuinely \
        describe the item — its kind, category, use, or where it is kept. \
        Example: for "cordless drill" reply: Tools, Power Tools, Garage, Hardware, DIY
        """
        return noThink ? "/no_think\n" + base : base
    }

    private func tagUser(name: String, description: String) -> String {
        var user = "Item: \(name)"
        let desc = description.trimmingCharacters(in: .whitespacesAndNewlines)
        if !desc.isEmpty { user += "\nDescription: \(desc)" }
        user += "\nTags:"
        return user
    }

    // MARK: - Reconciliation (ported from Android)

    private func reconcile(raw: String, existing: [HBTag]) -> TagSuggestions {
        var existingBySingular: [String: HBTag] = [:]
        for t in existing { existingBySingular[singular(t.name)] = t }

        var matched: [String] = []
        var novel: [String] = []
        var seen = Set<String>()

        for tag in parse(raw) {
            let key = singular(tag)
            guard !key.isEmpty, !seen.contains(key) else { continue }
            seen.insert(key)
            if let hit = existingBySingular[key] {
                if !matched.contains(hit.id) { matched.append(hit.id) }
            } else {
                novel.append(titleCase(tag))
            }
        }
        return TagSuggestions(matchedIds: matched, novel: novel)
    }

    private func parse(_ raw: String) -> [String] {
        var text = stripThinking(raw)
        if let r = text.range(of: "Tags:", options: .caseInsensitive) {
            text = String(text[r.upperBound...])
        }
        let line = text.split(whereSeparator: { $0.isNewline }).first.map(String.init) ?? text
        let trimChars = CharacterSet(charactersIn: " -*.\"'#\t")
        var result: [String] = []
        for piece in line.split(whereSeparator: { $0 == "," || $0 == ";" }) {
            let t = piece.trimmingCharacters(in: trimChars)
            let words = t.split(separator: " ")
            guard t.count >= 2, t.count <= 30, words.count <= 3 else { continue }
            if result.contains(where: { $0.caseInsensitiveCompare(t) == .orderedSame }) { continue }
            result.append(t)
            if result.count >= 6 { break }
        }
        return result
    }

    /// Strip `<think>...</think>` blocks (Qwen3 may leak them even with /no_think).
    private func stripThinking(_ raw: String) -> String {
        var s = raw
        while let start = s.range(of: "<think>") {
            if let end = s.range(of: "</think>", range: start.upperBound..<s.endIndex) {
                s.removeSubrange(start.lowerBound..<end.upperBound)
            } else {
                s.removeSubrange(start.lowerBound..<s.endIndex)
                break
            }
        }
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Lowercase + collapse spaces + fold simple plurals so "Spices" matches "spice".
    private func singular(_ s: String) -> String {
        var w = s.lowercased().trimmingCharacters(in: .whitespaces)
        w = w.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        if w.hasSuffix("ies"), w.count > 3 { return String(w.dropLast(3)) + "y" }
        // Skip sibilant -es (boxes, glasses) — rare in inventory tags.
        if w.hasSuffix("s"), !w.hasSuffix("ss"), !w.hasSuffix("es"), w.count > 1 {
            return String(w.dropLast())
        }
        return w
    }

    /// Title-case only all-lowercase words, preserving deliberate casing (USB, iPhone).
    private func titleCase(_ s: String) -> String {
        s.split(separator: " ").map { word -> String in
            let str = String(word)
            if !str.isEmpty, !str.contains(where: { $0.isUppercase }) {
                return str.prefix(1).uppercased() + String(str.dropFirst())
            }
            return str
        }
        .joined(separator: " ")
    }
}
